File: inst/d3/scatterplot/gaugechart.js
const val = Math.min(Math.max(data, options.min), options.max);
const minVal = options.min;
const maxVal = options.max;
const title = options.title || "";
const warningZone = options.warningZone;
const warningColor = options.warningColor || "orange";
const dangerZone = options.dangerZone;
const dangerColor = options.dangerColor || "red";
const arcBgColor = options.arcBgColor || "#ddd";
const needleColor = options.needleColor || "crimson";
const pivotColor = options.pivotColor || "steelblue";
const titleFontSize = options.titleFontSize || 16;
const valueFontSize = options.valueFontSize || 14;
const fontFamily = options.font || "Verdana, Geneva, Tahoma, sans-serif";
const bgcol = options.bgcol || "white";
// Needle animation (default on): the needle sweeps from the minimum up to the
// value on load and settles with a small wobble, like a car dashboard.
const animate = options.animate === undefined ? true : options.animate;
const animationDuration = options.animationDuration || 1500;

const cx = width / 2;
const cy = height / 2;
const radius = Math.min(width, height) / 2 * 0.82;
const arcWidth = radius * 0.16;

// Gauge spans 240 degrees: from -120° to +120° measured clockwise from north (12 o'clock)
// In D3 arc convention (0 = north, clockwise positive):
const startAngle = -(2 * Math.PI / 3);  // -120° = 8 o'clock
const endAngle   =  (2 * Math.PI / 3);  // +120° = 4 o'clock
const totalSpan  = endAngle - startAngle;

// Background SVG fill
svg.style("background-color", bgcol);

// Outer decorative circle (gives the bezel look)
const bezelGradId = "bezelGrad";
const bezelGrad = svg.append("defs").append("radialGradient")
  .attr("id", bezelGradId)
  .attr("cx", "40%")
  .attr("cy", "35%")
  .attr("r", "60%");
bezelGrad.append("stop").attr("offset", "0%").attr("stop-color", "#f5f5f5");
bezelGrad.append("stop").attr("offset", "70%").attr("stop-color", "#d0d0d0");
bezelGrad.append("stop").attr("offset", "100%").attr("stop-color", "#b0b0b0");

svg.append("circle")
  .attr("cx", cx)
  .attr("cy", cy)
  .attr("r", radius + arcWidth * 0.8)
  .attr("fill", `url(#${bezelGradId})`)
  .attr("stroke", "#999")
  .attr("stroke-width", 1.5);

// Inner face circle
svg.append("circle")
  .attr("cx", cx)
  .attr("cy", cy)
  .attr("r", radius - arcWidth * 0.15)
  .attr("fill", "#fafafa");

// Arc background track
const arcGen = d3.arc()
  .innerRadius(radius - arcWidth)
  .outerRadius(radius)
  .startAngle(startAngle)
  .endAngle(endAngle);

svg.append("path")
  .attr("transform", `translate(${cx},${cy})`)
  .attr("d", arcGen())
  .attr("fill", arcBgColor);

// Warning zone arc (e.g. orange)
const warningFraction = (warningZone - minVal) / (maxVal - minVal);
const warningStartAngle = startAngle + warningFraction * totalSpan;

const dangerFraction = (dangerZone - minVal) / (maxVal - minVal);
const dangerStartAngle = startAngle + dangerFraction * totalSpan;

if (warningFraction < 1) {
  // Clamp warning zone to end where danger zone begins
  const warningEndAngle = dangerFraction < 1 ? dangerStartAngle : endAngle;

  const warningArcGen = d3.arc()
    .innerRadius(radius - arcWidth)
    .outerRadius(radius)
    .startAngle(warningStartAngle)
    .endAngle(warningEndAngle);

  svg.append("path")
    .attr("transform", `translate(${cx},${cy})`)
    .attr("d", warningArcGen())
    .attr("fill", warningColor);
}

// Danger zone arc (e.g. red)
if (dangerFraction < 1) {
  const dangerArcGen = d3.arc()
    .innerRadius(radius - arcWidth)
    .outerRadius(radius)
    .startAngle(dangerStartAngle)
    .endAngle(endAngle);

  svg.append("path")
    .attr("transform", `translate(${cx},${cy})`)
    .attr("d", dangerArcGen())
    .attr("fill", dangerColor);
}

// Tick marks — 11 major ticks (one per 10% of range)
const numTicks = 11;
const tickInnerR = radius - arcWidth - 2;
const tickOuterR = radius - arcWidth - 10;

for (let i = 0; i < numTicks; i++) {
  const fraction = i / (numTicks - 1);
  const angle = startAngle + fraction * totalSpan;
  const sinA = Math.sin(angle);
  const cosA = Math.cos(angle);

  svg.append("line")
    .attr("x1", cx + tickInnerR * sinA)
    .attr("y1", cy - tickInnerR * cosA)
    .attr("x2", cx + tickOuterR * sinA)
    .attr("y2", cy - tickOuterR * cosA)
    .attr("stroke", "#555")
    .attr("stroke-width", i === 0 || i === numTicks - 1 ? 2 : 1.5);
}

// Minor ticks — 4 per major interval
const numMinorTicks = (numTicks - 1) * 4;
const minorTickInnerR = radius - arcWidth - 2;
const minorTickOuterR = radius - arcWidth - 6;

for (let i = 0; i <= numMinorTicks; i++) {
  if (i % 4 === 0) continue; // skip positions that coincide with major ticks
  const fraction = i / numMinorTicks;
  const angle = startAngle + fraction * totalSpan;
  const sinA = Math.sin(angle);
  const cosA = Math.cos(angle);

  svg.append("line")
    .attr("x1", cx + minorTickInnerR * sinA)
    .attr("y1", cy - minorTickInnerR * cosA)
    .attr("x2", cx + minorTickOuterR * sinA)
    .attr("y2", cy - minorTickOuterR * cosA)
    .attr("stroke", "#888")
    .attr("stroke-width", 1);
}

// Min / max value labels
const labelRadius = radius - arcWidth - 30;
const startLabelAngle = startAngle;
const endLabelAngle = endAngle;

svg.append("text")
  .attr("x", cx + labelRadius * Math.sin(startLabelAngle))
  .attr("y", cy - labelRadius * Math.cos(startLabelAngle) + 4)
  .attr("text-anchor", "middle")
  .attr("font-size", Math.max(10, radius * 0.1))
  .attr("fill", "#555")
  .attr("font-family", fontFamily)
  .text(minVal);

svg.append("text")
  .attr("x", cx + labelRadius * Math.sin(endLabelAngle))
  .attr("y", cy - labelRadius * Math.cos(endLabelAngle) + 4)
  .attr("text-anchor", "middle")
  .attr("font-size", Math.max(10, radius * 0.1))
  .attr("fill", "#555")
  .attr("font-family", fontFamily)
  .text(maxVal);

// Needle — drawn pointing straight up (north) inside a group centred on the
// pivot, then rotated to the target angle. Rotating a group (rather than
// recomputing the polygon) lets us animate the sweep smoothly.
const valueFraction = (val - minVal) / (maxVal - minVal);
const needleAngle = startAngle + valueFraction * totalSpan;   // radians
const needleAngleDeg = needleAngle * 180 / Math.PI;
const startAngleDeg  = startAngle  * 180 / Math.PI;
const endAngleDeg    = endAngle    * 180 / Math.PI;

const needleLength = radius - arcWidth - 12;
const needleTailLength = radius * 0.12;
const needleHalfBase = Math.max(4, radius * 0.04);

// Polygon pointing up: tip → right base → tail → left base
const needlePoints = [
  `0,${-needleLength}`,
  `${needleHalfBase},0`,
  `0,${needleTailLength}`,
  `${-needleHalfBase},0`
].join(" ");

const needleGroup = svg.append("g")
  .attr("transform",
    `translate(${cx},${cy}) rotate(${animate ? startAngleDeg : needleAngleDeg})`);

needleGroup.append("polygon")
  .attr("points", needlePoints)
  .attr("fill", needleColor);

// Sweep from the minimum up to the value, overshooting slightly and settling
// back — the way an analog car needle jumps and wobbles into place. The angle
// is clamped so the wobble never swings visibly past the ends of the gauge.
if (animate) {
  const clampLo = startAngleDeg - 6;
  const clampHi = endAngleDeg + 6;
  needleGroup.transition()
    .duration(animationDuration)
    .ease(d3.easeElasticOut.amplitude(1).period(0.6))
    .attrTween("transform", () => {
      const interp = d3.interpolate(startAngleDeg, needleAngleDeg);
      return t => {
        const a = Math.max(clampLo, Math.min(clampHi, interp(t)));
        return `translate(${cx},${cy}) rotate(${a})`;
      };
    });
}

// Center pivot circle
svg.append("circle")
  .attr("cx", cx)
  .attr("cy", cy)
  .attr("r", Math.max(6, radius * 0.07))
  .attr("fill", pivotColor)
  .attr("stroke", "white")
  .attr("stroke-width", 2);

// Title text (upper center of gauge face)
if (title) {
  svg.append("text")
    .attr("x", cx)
    .attr("y", cy - radius * 0.28)
    .attr("text-anchor", "middle")
    .attr("font-size", titleFontSize)
    .attr("fill", "#333")
    .attr("font-family", fontFamily)
    .attr("font-weight", "bold")
    .text(title);
}

// Current value text (below center)
svg.append("text")
  .attr("x", cx)
  .attr("y", cy + radius * 0.52)
  .attr("text-anchor", "middle")
  .attr("font-size", valueFontSize)
  .attr("fill", "#333")
  .attr("font-family", fontFamily)
  .text(val);

File: R/gauge_chart.R
#' Create a gauge (speedometer) chart.
#'
#' @description
#' A gauge (speedometer) chart displays a single quantitative measure
#' against colored zones (warning and danger) with a needle indicator.
#' By default the needle animates on load, sweeping up from the minimum to
#' the value and settling with a small wobble, like a car dashboard.
#'
#' @param value Numeric. The current value to display on the gauge.
#' @param min Numeric. The minimum value of the gauge. Defaults to 0.
#' @param max Numeric. The maximum value of the gauge. Defaults to 100.
#' @param title Character. Label shown inside the gauge. Defaults to "".
#' @param warningZone Numeric. The value at which the warning (first) arc begins.
#' Defaults to 75\% of max.
#' @param warningColor The color of the warning zone arc. Defaults to "orange".
#' @param dangerZone Numeric. The value at which the danger (second, critical) arc begins.
#' Defaults to 90\% of max.
#' @param dangerColor The color of the danger zone arc. Defaults to "red".
#' @param arcBgColor The background color of the gauge arc track. Defaults to "#ddd".
#' @param needleColor The color of the needle. Defaults to "crimson".
#' @param pivotColor The color of the center pivot circle. Defaults to "steelblue".
#' @param titleFontSize The font size of the gauge title. Defaults to 16.
#' @param valueFontSize The font size of the current value label. Defaults to 14.
#' @param font The font family used for all text. Defaults to
#' "Verdana, Geneva, Tahoma, sans-serif".
#' @param bgcol The background color of the SVG. Defaults to "white".
#' @param animate Logical. When \code{TRUE} (default), the needle sweeps from
#' the minimum up to the value on load and settles with a small wobble. Set to
#' \code{FALSE} to draw the needle at the value immediately with no animation.
#' @param animationDuration Numeric. Duration of the needle sweep in
#' milliseconds. Defaults to 1500.
#' @param width Optional. The width of the SVG output.
#' @param height Optional. The height of the SVG output.
#'
#' @return An SVG gauge chart.
#' @export
#'
#' @examples
#' # Basic usage
#' gauge_chart(value = 8, title = "Memory")
#'
#' # Custom warning and danger thresholds
#' gauge_chart(
#'   value = 72,
#'   min = 0,
#'   max = 100,
#'   title = "CPU Load",
#'   warningZone  = 60,
#'   warningColor = "orange",
#'   dangerZone   = 80,
#'   dangerColor  = "red"
#' )
#'
#' # Non-percentage range (response time in ms)
#' gauge_chart(
#'   value = 530,
#'   min = 0,
#'   max = 1000,
#'   title = "Response (ms)",
#'   warningZone  = 400,
#'   dangerZone   = 700
#' )
gauge_chart <- function(
    value,
    min = 0,
    max = 100,
    title = "",
    warningZone = NULL,
    warningColor = "orange",
    dangerZone = NULL,
    dangerColor = "red",
    arcBgColor = "#ddd",
    needleColor = "crimson",
    pivotColor = "steelblue",
    titleFontSize = 16,
    valueFontSize = 14,
    font = "Verdana, Geneva, Tahoma, sans-serif",
    bgcol = "white",
    animate = TRUE,
    animationDuration = 1500,
    width = NULL,
    height = NULL
) {

  if (!is.numeric(value) || length(value) != 1) {
    stop("'value' must be a single numeric value.")
  }

  if (min >= max) {
    stop("'min' must be less than 'max'.")
  }

  if (grepl(";", font)) {
    stop("please remove the ';' character from your font argument")
  }

  if (is.null(warningZone)) {
    warningZone <- min + 0.75 * (max - min)
  }

  if (is.null(dangerZone)) {
    dangerZone <- min + 0.90 * (max - min)
  }

  r2d3::r2d3(
    data = value,
    script = system.file("d3/scatterplot/gaugechart.js", package = "ddplot"),
    options = list(
      min = min,
      max = max,
      title = title,
      warningZone = warningZone,
      warningColor = warningColor,
      dangerZone = dangerZone,
      dangerColor = dangerColor,
      arcBgColor = arcBgColor,
      needleColor = needleColor,
      pivotColor = pivotColor,
      titleFontSize = titleFontSize,
      valueFontSize = valueFontSize,
      font = font,
      bgcol = bgcol,
      animate = animate,
      animationDuration = animationDuration
    ),
    width = width,
    height = height
  )
}
